dogechain-lab/dogechain
# [HardFork] Deposit block reward to validator after detroit hard fork

# Description

The design and implementation of the POS is heavily inspired by the Binance BSC chain, which has been running stably for over 2 years.
We take the concepts of `deposit` and `slash`, combined with our IBFT implementation, and strike a balance between efficiency and security.

Most important modifications include:
* Create a new `detroit` hard fork.
  * Upgrade `ValidatorSet` and `Bridge` contract byte code and initial storage.
* Validator will include a deposit transaction in each block. The transaction:
  * not broadcast,
  * marked as system transaction,
  * verified by all validators,
  * included at last so as to include other transactions sent by validators.
* Validator will include a slash transaction when some need to be punished. The transaction:
  * not broadcast,
  * marked as system transaction,
  * verified by all validators,
  * only slash the first validator who is the culprit,
  * included at the penultimate.
* Verify block timestamp after hard fork.
  * This will reduce the threat of block execution.
* Reduce `IBFT` waiting timeout.
  * The previous implementation used an exponentially growing timeout and the network seemed to stop after a few rounds.
  * Validators don't need too long time a timeout under an expected network topology, so we correlate the timeout period and the change in the number of validator numbers.
  * Occasionally, the network does fail, causing the vote to fail all the time. So we'd better specify a maximum round to ensure that all validators' views eventually agree.
* Hide transaction Hash field.
  * Which is buggy when we forget to compute its hash and set it to the transaction.
  * Use cache method like `go-ethereum` implementation, follow Law of Demeter (also known as The Least Knowledge Principle).

# Changes include


# Breaking changes

A hard fork is adopting, we need to upgrade all nodes before hard fork block height.

# Checklist


## Testing


### Manual tests

* Set up a 4 validators and 1 rpc-node network.
  * example [genesis.json](https://github.com/dogechain-lab/dogechain/files/9816654/genesis.txt)
* Deploy all reference [contracts](https://github.com/dogechain-lab/contracts/commit/8ae6cf37fb00604ca2c56d5d8baeb4e44a28727a).
* Benchmark it for 1-2 epochs.

You should see the network works fine, and delegator who staked on nodes gain token reward after the stake expires.
